tools/composefs-info.c

```diff
@@ -0,0 +1,294 @@
+	print_escaped(*path == 0 ? "/" : path, -1, ESCAPE_STANDARD);
+	printf(" %" PRIu64 " %s%o %u %u %u %u %" PRIi64 ".%u ",
+	       lcfs_node_get_size(target), target == node ? "" : "@",
+	       lcfs_node_get_mode(target), lcfs_node_get_nlink(target),
+	       lcfs_node_get_uid(target), lcfs_node_get_gid(target),
+	       lcfs_node_get_rdev(target), (int64_t)mtime.tv_sec,
+	       (unsigned int)mtime.tv_nsec);
+	print_escaped(payload ? payload : "-", -1, ESCAPE_LONE_DASH);
+
+	if (digest) {
+		char digest_str[LCFS_DIGEST_SIZE * 2 + 1] = { 0 };
+		digest_to_string(digest, digest_str);
+		printf(" %s", digest_str);
+	} else {
+		printf(" -");
+	}
+
+	size_t n_xattr = lcfs_node_get_n_xattr(target);
+	for (size_t i = 0; i < n_xattr; i++) {
+		const char *name = lcfs_node_get_xattr_name(target, i);
+		size_t value_len;
+		const char *value = lcfs_node_get_xattr(target, name, &value_len);
+
+		printf(" ");
+		print_escaped(name, -1, ESCAPE_EQUAL);
+		printf("=");
+		print_escaped(value, value_len, ESCAPE_EQUAL);
+	}
+
+	printf("\n");
+
+	for (size_t i = 0; i < lcfs_node_get_n_children(node); i++) {
+		struct lcfs_node_s *child = lcfs_node_get_child(node, i);
+		cleanup_free char *child_path = NULL;
+
+		asprintf(&child_path, "%s/%s", path, lcfs_node_get_name(child));
+
+		dump_node(child, child_path);
+	}
+}
+
+static void get_objects(struct lcfs_node_s *node, Hash_table *ht)
+{
+	uint32_t mode = lcfs_node_get_mode(node);
+	uint32_t type = mode & S_IFMT;
+	const char *payload = lcfs_node_get_payload(node);
+
+	if (type == S_IFREG && payload) {
+		if (hash_insert(ht, payload) == NULL) {
+			errx(EXIT_FAILURE, "Out of memory");
+		}
+	}
+
+	for (size_t i = 0; i < lcfs_node_get_n_children(node); i++) {
+		struct lcfs_node_s *child = lcfs_node_get_child(node, i);
+		get_objects(child, ht);
+	}
+}
+
+static size_t str_ht_hash(const void *entry, size_t table_size)
+{
+	return hash_string(entry, table_size);
+}
+
+static bool str_ht_eq(const void *entry1, const void *entry2)
+{
+	return strcmp(entry1, entry2) == 0;
+}
+
+static int cmp_obj(const void *_a, const void *_b)
+{
+	const char *const *a = _a;
+	const char *const *b = _b;
+	return strcmp(*a, *b);
+}
+
+static void print_objects(struct lcfs_node_s *node)
+{
+	Hash_table *ht = hash_initialize(0, NULL, str_ht_hash, str_ht_eq, NULL);
+	if (ht == NULL)
+		errx(EXIT_FAILURE, "Out of memory");
+
+	get_objects(node, ht);
+
+	size_t n_objects = hash_get_n_entries(ht);
+	char **objects = calloc(n_objects, sizeof(char *));
+
+	hash_get_entries(ht, (void **)objects, n_objects);
+
+	qsort(objects, n_objects, sizeof(char *), cmp_obj);
+
+	for (size_t i = 0; i < n_objects; i++)
+		printf("%s\n", objects[i]);
+
+	hash_free(ht);
+}
+
+static void usage(const char *argv0)
+{
+	fprintf(stderr, "usage: %s [ls|objects|dump] IMAGE\n", argv0);
+}
+
+int main(int argc, char **argv)
+{
+	const char *bin = argv[0];
+	int fd;
+	cleanup_node struct lcfs_node_s *root = NULL;
+	const char *image_path = NULL;
+	const char *command;
```

Comment:
So not really related to this exactly but...can I arm twist you into using declare-and-initialize style in this codebase?  I find it much more readable than "big dump of variables at the top of the function".  It plays more nicely with `__attribute__(cleanup)` too because there's no footgun around forgetting to init to `NULL` etc.

Reply:
sure